v1.9 IPSec key rotation without agent restart

jibi opened the pull request:
This is the v1.9 backport of https://github.com/cilium/cilium/pull/19814.

Just some minor conflicts on `nodediscovery: split StartDiscovery into smaller functions` as the `v1.9` branch does not include the changes to make `NodeDiscovery.LocalNode` a private field

nebril committed:
ci: use python3 instead of python

[ upstream commit dddbbe709e2827873420fef9b635152340f37f91 ]

Our CI nodes no longer have `python` binary, python3 is available
instead.

Signed-off-by: Maciej Kwiek <maciej@isovalent.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

gandro committed:
health: Move endpoint IP to node package

[ upstream commit ed934cb958c52515be37998038301229959a0fd2 ]

The Cilium health endpoint IP is the only static local node field which
is not accessible from the node package. To avoid code modifying the
`NodeDiscovery.LocalNode` field (which will become more dynamic in a
subsequent mode), the health IP is now moved to the node package with a
corresponding getter and setter like all other static node fields.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

gandro committed:
nodediscovery: Make LocalNode object private

[ upstream commit e52fe1d59d1c6f51a76993641ea9412e0f44f749 ]

The LocalNode currently consists of static fields which are never
changed after the local node has been initialized. However, with the
introduction of secondary allocation CIDRs, we want to be update the
secondary allocation CIDR fields dynamically at run-time.

To ensure writes to the object are always protected by a mutex, and that
to be able to ensure that any changes to the object are propagated to
relevant subsystems, this commit makes the `LocalNode` object private.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
nodediscovery: use time.After() instead of time.NewTimer().C

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
nodediscovery: split StartDiscovery into smaller functions

in preparation to a subsequent change, this commit moves some of the
(*NodeDiscovery).StartDiscovery() logic into separated functions:

* fillLocalNode(), responsible for syncing the localNode object with the
  actual state of the node
* updateLocalNode(), responsible for publishing the updated KV store
  entry and/or CiliumNode object for the local node

In addition to that, this commit introduces also the UpdateLocalNode()
function, which ensure that the proper lock is held while calling
updateLocalNode(), and allows optionally to sync the localNode object
state by calling fillLocalNode().

No functional changes are introduced.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
Move fswatcher package out of crypto/certloader

Given that this logic is totally agnostic to the crypto/certloader
package and in a subsequent commit it will be used also in the IPSec
subsystem, make it a top level package.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

pchaigno commented on daemon/cmd/daemon.go:
```diff
@@ -690,7 +690,7 @@ func NewDaemon(ctx context.Context, epMgr *endpointmanager.EndpointManager, dp d
 
 	// Must occur after d.allocateIPs(), see GH-14245 and its fix.
-	d.nodeDiscovery.StartDiscovery(nodeTypes.GetName())
+	d.nodeDiscovery.StartDiscovery()
```
Could we backport 9c1cb4856322a80174799dae142ef761e51e2310 instead of cherry-picking a subset of its changes?

pchaigno reviewed (commented):
I see conflicts reported by GitHub. Is that expected? If it was based on some branch than v1.9, then maybe it could be opened against that branch (e.g., `v1.9.x`)? EDIT: GitHub doesn't allow that :disappointed: 

Other than the comment below and other comments in the upstream PR, looks good to me. Nothing blocking.

gandro commented on pkg/nodediscovery/nodediscovery.go:
```diff
@@ -275,6 +289,14 @@ func (n *NodeDiscovery) StartDiscovery(nodeName string) {
 	}
 }
 
+func (n *NodeDiscovery) UpdateLocalNode(fillLocalNode bool) {
```
`fillLocalNode` is always true. Why is it needed?

gandro commented on pkg/nodediscovery/nodediscovery.go:
```diff
@@ -172,8 +175,44 @@ func (n *NodeDiscovery) JoinCluster(nodeName string) {
-func (n *NodeDiscovery) StartDiscovery(nodeName string) {
-	n.LocalNode.Name = nodeName
+func (n *NodeDiscovery) StartDiscovery() {
+	n.localNodeLock.Lock()
+	defer n.localNodeLock.Unlock()
```
Is this mutex needed on v1.9? `StartDiscovery` cannot be called more than once anyway and it doesn't seem to be used anywhere else.

On v1.9, we also cannot really protect writes to `LocalNode`, as it is exported.

gandro replied:
Oh, I missed the fact that the watcher is calling `fillLocalNode` indirectly. I think that might be inherently racy, as the fields could be written to while some other code is reading it.

gandro reviewed (commented):
The fsnotify code itself looks good to me. A few comments around the node discovery code though.

jibi committed:
ipsec: implement keyfile watcher

This commit adds a new filesystem watcher for the IPSec subsystem, which
in turn will enable automatic key rotation without the need to restart
the agent whenever a change to the IPSec keyfile is detected.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
node: add GetAbsoluteNodeName method

This commit backport just the GetAbsoluteNodeName method from
3203df90821f562d655bbda24fd87716f6c7661a, as it will be required in a
subsequent commit.

Signed-off-by: Maddy007-maha <mahadev.panchal@accuknox.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
nodediscovery: Unify source of truth for node name

[ upstream commit 9c1cb4856322a80174799dae142ef761e51e2310 ]

Clustermesh related code is relying on
`NodeDiscovery.LocalNode.FullName()` to obtain the globally unique name
of the local node. This field is initialized from `nodeTypes.GetName()`.
This code removes all uses of `LocalNode.FullName()` and replaces it
with an equivalent call to `nodeTypes.GetAbsoluteNodeName()`. This
allows us to make the `NodeDiscovery.LocalNode` private in a subsequent
commit.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

gandro reviewed (approved):
Neat, this looks good to me now also in terms of the node discovery logic!

jibi committed:
ipsec: fix stale keys reclaim logic

The current logic used to reclaim stale IPSec keys is incorrect. In
fact, whenever a new key is added we just spawn a new goroutine which
after linux_defaults.IPsecKeyDeleteDelay time will delete all keys
different than the one just added.

The issue with this approach is that if a new key is added while one of
these goroutine is still running, this goroutine will end up deleting
the new key.

This commit fixes this by introducing a single new goroutine that
periodically removes any stale key (XFRM state and XFRM policy) after
enough time has passed since the current key has been added.

Fixes: #19814
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
ipsec: rename ipSecKeysGlobalLock to ipSecLock

as now this lock is used to protect also ipSecCurrentKeySPI in addition
to ipSecKeysGlobal.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
ipsec: add ipSecXfrmMark{Set,Get}SPI helpers

which allows to encode and extract an SPI value in and from a XfrmMark

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
logfields: add SPI and OldSPI IPSec constants

Signed-off-by: Gilberto Bertin <jibi@cilium.io>
